casper/src/main/scala/coop/rchain/casper/ReportingCasper.scala

```diff
@@ -295,43 +289,46 @@ object ReportingCasper {
-                           runtime.reportingSpace.getReport.map(_.asRight[ReplayFailure])
-                         case Left(ex: ReplayException) => {
-                           Log[F]
-                             .error(s"Failed during processing of deploy: ${processedDeploy}") >>
-                             (none[DeployData], UnusedCommEvent(ex): Failed)
-                               .asLeft[Seq[ReportingEvent]]
-                               .pure[F]
-                         }
-                         case Left(ex) =>
-                           (none[DeployData], UserErrors(Vector(ex)): Failed)
-                             .asLeft[Seq[ReportingEvent]]
-                             .pure[F]
-                       }
-                   }
-               }
-      } yield cont
+        replayEvaluateResultEither <- computeEffect(runtime, runtime.replayReducer)(
+                                       processedDeploy.deploy
+                                     ).attempt
+        failureEither <- replayEvaluateResultEither match {
+                          case Right(EvaluateResult(replayCost, replayUserErrors)) =>
+                            if (isFailed != replayUserErrors.nonEmpty)
+                              ReplayFailure
+                                .replayStatusMismatch(isFailed, replayUserErrors.nonEmpty)
+                                .asLeft[Seq[ReportingEvent]]
+                                .pure[F]
+                            else if (replayUserErrors.nonEmpty)
+                              runtime.reportingSpace.revertToSoftCheckpoint(softCheckpoint) >> runtime.reportingSpace.getReport
+                                .map(_.asRight[ReplayFailure])
+                            else if (cost.cost != replayCost.value)
+                              ReplayFailure
+                                .replayCostMismatch(deploy, cost.cost, replayCost.value)
+                                .asLeft[Seq[ReportingEvent]]
+                                .pure[F]
+                            else
+                              runtime.reportingSpace.checkReplayData().attempt >>= {
+                                case Right(_) =>
+                                  runtime.reportingSpace.getReport.map(_.asRight[ReplayFailure])
+                                case Left(replayException: ReplayException) =>
+                                  ReplayFailure
+                                    .unusedCOMMEvent(deploy, replayException)
+                                    .asLeft[Seq[ReportingEvent]]
+                                    .pure[F]
+                                case Left(throwable) =>
+                                  ReplayFailure
+                                    .internalError(deploy, throwable)
+                                    .asLeft[Seq[ReportingEvent]]
+                                    .pure[F]
+                              }
+                          case Left(throwable) =>
```

Comment:
Just please make sure it does everything the original one did :smile:

Reply:
IMO it has the following benefits:
1. eliminates duplication in error handling
1. has much cleaner "happy-path"
1. you can actually read what it does (well, minus our formatting style that was apparently chosen by a person programming on teletype (maxColumn = 100, really?)

YMMV

Reply:
Thanks! One important difference between your code and mine is that yours compares costs before checking for user errors and mine does the opposite. We can only compare costs if there are no errors since costs become non-deterministic in the event of errors. At any rate, this is very helpful.

Reply:
Ah, I see it. Interesting. Do you want me to attempt (hehe, pun intended) to fix the example?

Reply:
So I have another suggestion. If costs are non-det in presence of failures than maybe it should be reflected in the type. Looks like simmple ADT will do (`data EvaluateResult = Correct cost | CatastropheOfBiblicalProportions errors` )

Reply:
> So I have another suggestion. If costs are non-det in presence of failures than maybe it should be reflected in the type. Looks like simmple ADT will do (data EvaluateResult = Correct cost | CatastropheOfBiblicalProportions errors )

I'll repeat what I wrote to you in Discord: 

Also, I've tried your ADT suggestion. I think it's unnecessary. Eventually (possibly/hopefully before main-net), we'll be able to compare program errors reliably. In that cost, it will make sense to return an ADT that includes cost and errors. Right now, it introduces unnecessary branching in RuntimeManager, since I would have to duplicate the check that failure status' match. Whichever ADT I get back during replay (one of two), I have to verify that I got the same one last time. I can appreciate the motivation though :thumbsup:

Reply:
Agreed on that

Reply:
> One important difference between your code and mine is that yours compares costs before checking for user errors and mine does the opposite. We can only compare costs if there are no errors since costs become non-deterministic in the event of errors. 

So it seems to me that this simple change will make it right:
> + result => !result.executionFailed && cost.cost != result.cost.value)
> - result => cost.cost != result.cost.value)

Costs are not compared, subsequent `semiflatMap` calls `revertToCheckpoint` and they lived happily everafter